File: 2nd/homework_5/id_69/99_RecoverBinarySearchTree.py
# https://leetcode.com/problems/recover-binary-search-tree/description/
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import sys


class Solution(object):

    def __init__(self):
        self.prev = TreeNode(-sys.maxsize - 1)
        self.first = None
        self.second = None

    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        self.in_order_traversal(root)
        if self.first is not None and self.second is not None:
            self.first.val, self.second.val = self.second.val, self.first.val

    def in_order_traversal(self, root):
        if root is None:
            return
        self.in_order_traversal(root.left)
        if self.first is None and root.val <= self.prev.val:
            self.first = self.prev
        if self.first is not None and root.val <= self.prev.val:
            self.second = root
        self.prev = root
        self.in_order_traversal(root.right)

File: 2nd/homework_5/id_45/99_recover_binary_search_tree.cpp
/**
 *  * Definition for a binary tree node.
 *  * struct TreeNode {
 *  *     int val;
 *  *     TreeNode *left;
 *  *     TreeNode *right;
 *  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *  * };
 *  */

class Solution {
public:
    TreeNode *n1 = NULL, *n2 = NULL;
    TreeNode *pre = new TreeNode(INT_MIN);

    void recoverTree(TreeNode* root) {
        if (!root)
            return;
        inorder(root);
        swap(n1->val, n2->val);
    }

    void inorder(TreeNode *root) {
        if (!root)
            return;
        inorder(root->left);
        if (!n1 && pre->val >= root->val)
            n1 = pre;
        if (n1 && pre->val >= root->val)
            n2 = root;
        pre = root;
        inorder(root->right);
    }
};
